File: client/src/main/scala/stasis/client/ops/recovery/RecoveryEntityKind.scala
package stasis.client.ops.recovery

import java.nio.file.Files

import scala.concurrent.ExecutionContext
import scala.concurrent.Future

import org.apache.pekko.Done
import org.apache.pekko.NotUsed
import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString

import stasis.client.analysis.Metadata
import stasis.client.analysis.PlatformMetadata
import stasis.client.collection.RecoveryCollector
import stasis.client.collection.RecoveryMetadataCollector
import stasis.client.collection.rules.SourceUri
import stasis.client.model.DatasetMetadata
import stasis.client.model.EntityMetadata
import stasis.client.model.EntityRef
import stasis.client.model.FilesystemMetadata
import stasis.client.model.TargetEntity
import stasis.client.ops.ParallelismConfig
import stasis.client.ops.recovery.stages.internal.DestagedByteStringSource

sealed trait RecoveryEntityKind {
  def collector(
    targetMetadata: DatasetMetadata,
    keep: (String, FilesystemMetadata.EntityState) => Boolean,
    destination: TargetEntity.Destination,
    providers: Providers,
    parallelism: ParallelismConfig
  )(implicit mat: Materializer): RecoveryCollector
}

object RecoveryEntityKind {
  trait Filesystem extends RecoveryEntityKind {
    def prepare(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers): Unit

    def write(entity: TargetEntity, ref: EntityRef.Filesystem, content: Source[ByteString, NotUsed], providers: Providers)(
      implicit mat: Materializer
    ): Future[Done]

    def applyMetadata(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers)(implicit
      ec: ExecutionContext
    ): Future[Done]
  }

  trait Library extends RecoveryEntityKind {
    def scheme: String

    def prepare(entity: TargetEntity, ref: EntityRef.Library, providers: Providers): Unit

    def write(entity: TargetEntity, ref: EntityRef.Library, content: Source[ByteString, NotUsed], providers: Providers)(implicit
      mat: Materializer
    ): Future[Done]

    def applyMetadata(entity: TargetEntity, ref: EntityRef.Library, providers: Providers)(implicit
      ec: ExecutionContext
    ): Future[Done]
  }

  def prepare(kinds: Seq[RecoveryEntityKind], entity: TargetEntity, providers: Providers): Unit =
    entity.destinationRef match {
      case ref: EntityRef.Filesystem =>
        kinds.collectFirst { case kind: Filesystem => kind.prepare(entity, ref, providers) }.getOrElse(())

      case ref: EntityRef.Library =>
        kinds
          .collectFirst { case kind: Library if kind.scheme == ref.scheme => kind.prepare(entity, ref, providers) }
          .getOrElse(())
    }

  def write(kinds: Seq[RecoveryEntityKind], entity: TargetEntity, content: Source[ByteString, NotUsed], providers: Providers)(
    implicit mat: Materializer
  ): Future[Done] =
    entity.destinationRef match {
      case ref: EntityRef.Filesystem =>
        kinds
          .collectFirst { case kind: Filesystem => kind.write(entity, ref, content, providers) }
          .getOrElse(Future.failed(new IllegalArgumentException("No filesystem recovery kind was registered")))

      case ref: EntityRef.Library =>
        kinds
          .collectFirst { case kind: Library if kind.scheme == ref.scheme => kind.write(entity, ref, content, providers) }
          .getOrElse(Future.failed(new IllegalArgumentException(s"No recovery kind was registered for scheme [${ref.scheme}]")))
    }

  def applyMetadata(kinds: Seq[RecoveryEntityKind], entity: TargetEntity, providers: Providers)(implicit
    ec: ExecutionContext
  ): Future[Done] =
    entity.destinationRef match {
      case ref: EntityRef.Filesystem =>
        kinds
          .collectFirst { case kind: Filesystem => kind.applyMetadata(entity, ref, providers) }
          .getOrElse(Future.failed(new IllegalArgumentException("No filesystem recovery kind was registered")))

      case ref: EntityRef.Library =>
        kinds
          .collectFirst { case kind: Library if kind.scheme == ref.scheme => kind.applyMetadata(entity, ref, providers) }
          .getOrElse(Future.failed(new IllegalArgumentException(s"No recovery kind was registered for scheme [${ref.scheme}]")))
    }

  object Filesystem extends Filesystem {
    override def collector(
      targetMetadata: DatasetMetadata,
      keep: (String, FilesystemMetadata.EntityState) => Boolean,
      destination: TargetEntity.Destination,
      providers: Providers,
      parallelism: ParallelismConfig
    )(implicit mat: Materializer): RecoveryCollector = {
      implicit val ec: ExecutionContext = mat.executionContext
      implicit val p: ParallelismConfig = parallelism

      new RecoveryCollector.Filesystem(
        targetMetadata = targetMetadata,
        keep = (entity, state) => keep(entity, state) && SourceUri.scheme(entity).isEmpty,
        destination = destination,
        metadataCollector = RecoveryMetadataCollector.Filesystem(checksum = providers.checksum),
        clients = providers.clients
      )
    }

    override def prepare(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers): Unit = {
      val directory = entity.existingMetadata match {
        case _: EntityMetadata.Directory   => ref.path
        case _: EntityMetadata.WithContent => ref.path.getParent
      }

      val _ = Files.createDirectories(
        directory,
        PlatformMetadata.forFileSystem(directory.getFileSystem).ownerOnlyDirectoryAttributes: _*
      )
    }

    override def write(
      entity: TargetEntity,
      ref: EntityRef.Filesystem,
      content: Source[ByteString, NotUsed],
      providers: Providers
    )(implicit mat: Materializer): Future[Done] = {
      implicit val prv: Providers = providers
      new DestagedByteStringSource(content).destage(to = ref.path)
    }

    override def applyMetadata(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers)(implicit
      ec: ExecutionContext
    ): Future[Done] =
      Metadata.applyEntityMetadataTo(metadata = entity.existingMetadata, entity = ref.path)(ec, providers.metadataDefaults)
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/recovery/RecoveryEntityKind.kt
package stasis.client_android.lib.ops.recovery

import okio.Source
import stasis.client_android.lib.analysis.Metadata
import stasis.client_android.lib.collection.FilesystemRecoveryCollector
import stasis.client_android.lib.collection.RecoveryCollector
import stasis.client_android.lib.collection.RecoveryMetadataCollector
import stasis.client_android.lib.collection.rules.SourceUri
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.EntityRef
import stasis.client_android.lib.model.FilesystemMetadata
import stasis.client_android.lib.model.TargetEntity
import stasis.client_android.lib.ops.recovery.stages.internal.DestagedByteStringSource.destage
import java.nio.file.Files
import java.nio.file.attribute.FileAttribute
import java.nio.file.attribute.PosixFilePermission
import java.nio.file.attribute.PosixFilePermissions

sealed interface RecoveryEntityKind {
    fun collector(
        targetMetadata: DatasetMetadata,
        keep: (String, FilesystemMetadata.EntityState) -> Boolean,
        destination: TargetEntity.Destination,
        providers: Providers
    ): RecoveryCollector

    interface Filesystem : RecoveryEntityKind {
        fun prepare(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers)

        suspend fun write(entity: TargetEntity, ref: EntityRef.Filesystem, content: Source, providers: Providers)

        suspend fun applyMetadata(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers)
    }

    interface Library : RecoveryEntityKind {
        val scheme: String

        fun prepare(entity: TargetEntity, ref: EntityRef.Library, providers: Providers)

        suspend fun write(entity: TargetEntity, ref: EntityRef.Library, content: Source, providers: Providers)

        suspend fun applyMetadata(entity: TargetEntity, ref: EntityRef.Library, providers: Providers)
    }

    companion object {
        fun prepare(kinds: List<RecoveryEntityKind>, entity: TargetEntity, providers: Providers) {
            when (val ref = entity.destinationRef) {
                is EntityRef.Filesystem ->
                    kinds.filterIsInstance<Filesystem>().firstOrNull()?.prepare(entity, ref, providers)

                is EntityRef.Library ->
                    kinds.filterIsInstance<Library>().firstOrNull { it.scheme == ref.scheme }
                        ?.prepare(entity, ref, providers)
            }
        }

        suspend fun write(
            kinds: List<RecoveryEntityKind>,
            entity: TargetEntity,
            content: Source,
            providers: Providers
        ) {
            when (val ref = entity.destinationRef) {
                is EntityRef.Filesystem ->
                    (kinds.filterIsInstance<Filesystem>().firstOrNull()
                        ?: throw IllegalArgumentException("No filesystem recovery kind was registered"))
                        .write(entity, ref, content, providers)

                is EntityRef.Library ->
                    (kinds.filterIsInstance<Library>().firstOrNull { it.scheme == ref.scheme }
                        ?: throw IllegalArgumentException("No recovery kind was registered for scheme [${ref.scheme}]"))
                        .write(entity, ref, content, providers)
            }
        }

        suspend fun applyMetadata(kinds: List<RecoveryEntityKind>, entity: TargetEntity, providers: Providers) {
            when (val ref = entity.destinationRef) {
                is EntityRef.Filesystem ->
                    (kinds.filterIsInstance<Filesystem>().firstOrNull()
                        ?: throw IllegalArgumentException("No filesystem recovery kind was registered"))
                        .applyMetadata(entity, ref, providers)

                is EntityRef.Library ->
                    (kinds.filterIsInstance<Library>().firstOrNull { it.scheme == ref.scheme }
                        ?: throw IllegalArgumentException("No recovery kind was registered for scheme [${ref.scheme}]"))
                        .applyMetadata(entity, ref, providers)
            }
        }

        val Filesystem: Filesystem = object : Filesystem {
            private val targetDirectoryAttributes: FileAttribute<Set<PosixFilePermission>> =
                PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString("rwx------"))

            override fun collector(
                targetMetadata: DatasetMetadata,
                keep: (String, FilesystemMetadata.EntityState) -> Boolean,
                destination: TargetEntity.Destination,
                providers: Providers
            ): RecoveryCollector =
                FilesystemRecoveryCollector(
                    targetMetadata = targetMetadata,
                    keep = { entity, state -> keep(entity, state) && SourceUri.scheme(entity) == null },
                    destination = destination,
                    metadataCollector = RecoveryMetadataCollector.Filesystem(checksum = providers.checksum),
                    clients = providers.clients
                )

            override fun prepare(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers) {
                val directory = when (entity.existingMetadata) {
                    is EntityMetadata.Directory -> ref.path
                    is EntityMetadata.File -> ref.path.parent
                    is EntityMetadata.Library -> ref.path.parent
                }

                Files.createDirectories(directory, targetDirectoryAttributes)
            }

            override suspend fun write(
                entity: TargetEntity,
                ref: EntityRef.Filesystem,
                content: Source,
                providers: Providers
            ) {
                content.destage(to = ref.path, providers = providers)
            }

            override suspend fun applyMetadata(entity: TargetEntity, ref: EntityRef.Filesystem, providers: Providers) {
                Metadata.applyEntityMetadataTo(metadata = entity.existingMetadata, entity = ref.path)
            }
        }
    }
}
